"null" time from CouchDB 2.1?

Intermittent in Travis:

    --- FAIL: TestCouch21 (11.29s)
    	test.go:403: Connecting to http://admin:abc123@localhost:6002/ ...
    	test.go:415: Connecting to http://localhost:6002/ ...
        --- FAIL: TestCouch21/Replicate (0.48s)
            --- FAIL: TestCouch21/Replicate/RW (0.48s)
                --- FAIL: TestCouch21/Replicate/RW/Admin (0.29s)
                    --- FAIL: TestCouch21/Replicate/RW/Admin/group (0.00s)
                        --- FAIL: TestCouch21/Replicate/RW/Admin/group/Cancel (0.02s)
                        	kt.go:258: Unexpected failure: 602/parsing time "null" as ""2006-01-02T15:04:05Z07:00"": cannot parse "null" as """
                        --- FAIL: TestCouch21/Replicate/RW/Admin/group/MissingTarget (0.03s)
                        	kt.go:258: Unexpected failure: 602/parsing time "null" as ""2006-01-02T15:04:05Z07:00"": cannot parse "null" as """
                        --- FAIL: TestCouch21/Replicate/RW/Admin/group/MissingSource (0.01s)
                        	kt.go:258: Unexpected failure: 602/parsing time "null" as ""2006-01-02T15:04:05Z07:00"": cannot parse "null" as """
                        --- FAIL: TestCouch21/Replicate/RW/Admin/group/ValidReplication (0.02s)
                        	kt.go:258: Unexpected failure: 602/parsing time "null" as ""2006-01-02T15:04:05Z07:00"": cannot parse "null" as """

--- a/scheduler17.go
+++ b/scheduler17.go
@@ -1,0 +1,58 @@
+// +build go1.7,!go1.8
+
+package couchdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+type schedulerDoc17 struct {
+	Database      string   `json:"database"`
+	DocID         string   `json:"doc_id"`
+	ReplicationID string   `json:"id"`
+	Source        string   `json:"source"`
+	Target        string   `json:"target"`
+	StartTime     nullTime `json:"start_time"`
+	LastUpdated   nullTime `json:"last_updated"`
+	State         string   `json:"state"`
+	Info          repInfo  `json:"info"`
+}
+
+type nullTime time.Time
+
+// inspired by https://github.com/golang/go/issues/9037#issuecomment-270932787
+func (t *nullTime) UnmarshalJSON(data []byte) error {
+	fmt.Printf("input = `%s`\n", string(data))
+	if string(data) == "null" {
+		return nil
+	}
+	var ts time.Time
+	if err := json.Unmarshal(data, &ts); err != nil {
+		return err
+	}
+	*t = nullTime(ts)
+	return nil
+}
+
+// UnmarshalJSON handles "null" values for times, which Go 1.7 doesn't natively
+// handle properly
+func (d *schedulerDoc) UnmarshalJSON(data []byte) error {
+	var doc schedulerDoc17
+	if err := json.Unmarshal(data, &doc); err != nil {
+		return err
+	}
+	*d = schedulerDoc{
+		Database:      doc.Database,
+		DocID:         doc.DocID,
+		ReplicationID: doc.ReplicationID,
+		Source:        doc.Source,
+		Target:        doc.Target,
+		StartTime:     time.Time(doc.StartTime),
+		LastUpdated:   time.Time(doc.LastUpdated),
+		State:         doc.State,
+		Info:          doc.Info,
+	}
+	return nil
+}

--- a/scheduler17_test.go
+++ b/scheduler17_test.go
@@ -1,0 +1,47 @@
+// +build go1.7,!go1.8
+
+package couchdb
+
+import (
+	"encoding/json"
+	"testing"
+	"time"
+
+	"github.com/flimzy/diff"
+	"github.com/flimzy/testy"
+)
+
+func TestNullTimeUnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    string
+		expected time.Time
+		err      string
+	}{
+		{
+			name:     "valid time",
+			input:    `"2017-11-17T19:56:09Z"`,
+			expected: parseTime(t, "2017-11-17T19:56:09Z"),
+		},
+		{
+			name:     "null",
+			input:    `null`,
+			expected: time.Time{},
+		},
+		{
+			name:  "invalid json",
+			input: `invalid json`,
+			err:   "invalid character 'i' looking for beginning of value",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var result nullTime
+			err := json.Unmarshal([]byte(test.input), &result)
+			testy.Error(t, test.err, err)
+			if d := diff.Interface(test.expected, time.Time(result)); d != nil {
+				t.Error(d)
+			}
+		})
+	}
+}

--- a/scheduler_test.go
+++ b/scheduler_test.go
@@ -6,6 +6,7 @@ import (
 	"errors"
 	"net/http"
 	"testing"
+	"time"
 
 	"github.com/flimzy/diff"
 	"github.com/flimzy/kivik"
@@ -547,6 +548,35 @@ func TestSRinnerUpdate(t *testing.T) {
 						reason: "db_not_found: could not open foo",
 					},
 				},
+			},
+		},
+		{
+			name: "null time",
+			r: &schedulerReplication{
+				database: "_replicator",
+				docID:    "56d257bd2125c8f15870b3ddd202c4ca",
+				db: newTestDB(&http.Response{
+					StatusCode: 200,
+					Header: http.Header{
+						"Server":         {"CouchDB/2.1.0 (Erlang OTP/17)"},
+						"Date":           {"Fri, 17 Nov 2017 13:05:52 GMT"},
+						"Content-Type":   {"application/json"},
+						"Content-Length": {"275"},
+						"Cache-Control":  {"must-revalidate"},
+					},
+					Body: Body(`{"database":"_replicator","doc_id":"733c70a35768b7a8fc2e178bd9003f1b","id":null,"source":"http://localhost:5984/kivik$replicate_rw_admin$5fbcf68d8d9aaee0/","target":"http://localhost:5984/foo/","state":null,"error_count":0,"info":null,"start_time":null,"last_updated":null}`),
+				}, nil),
+			},
+			expected: &schedulerReplication{
+				docID:         "733c70a35768b7a8fc2e178bd9003f1b",
+				database:      "_replicator",
+				replicationID: "",
+				source:        "http://localhost:5984/kivik$replicate_rw_admin$5fbcf68d8d9aaee0/",
+				target:        "http://localhost:5984/foo/",
+				startTime:     time.Time{},
+				lastUpdated:   time.Time{},
+				state:         "",
+				info:          repInfo{},
 			},
 		},
 	}

